gum/backend-arm64/gumstalker-arm64.c

```diff
@@ -808,6 +834,218 @@ gum_stalker_init (GumStalker * self)
+    _Unwind_SetIP (context, GPOINTER_TO_SIZE (resume_ip));
+
+    ctx->pending_calls--;
+  }
+
+  return reason;
+}
+
+static const void *
+gum_stalker_exception_find_fde (const void * pc,
+                                struct dwarf_eh_bases * bases)
+{
+  const void * result;
+  GumExecCtx * ctx;
+  gpointer real_address;
+
+  /*
+   * This function is responsible for finding the Frame Descriptor Entry
+   * associated with a given exception. To do this, it is provided with the PC
+   * of the entry to find.
+   *
+   * The PC provided is the address of the last byte of the instruction which
+   * called __cxa_throw. Since we store the address of the next instruction in
+   * our hashtable, (this is used by the personality function) we need to add 1
+   * to the value provided before we perform the lookup and subsequently
+   * subtract one from the value retrieved.
+   *
+   * If an exception is thrown whilst there is a code (rather than real) address
+   * in our stack (to allow us to execute excluded ranges), then the translation
+   * from code address to real address should have been inserted into the
+   * hashtable when the instrumented code was written. The _Unwind_Find_FDE
+   * function will be called with a code address (rather than the real address)
+   * which would usually fail, when this happends we need to translate the
+   * address before the call. If we have no associated entry in our lookup, then
+   * we can let the call proceed as normal.
+   */
+
+  ctx = gum_stalker_get_exec_ctx (GUM_IC_GET_REPLACEMENT_DATA (gum_interceptor_get_current_invocation (), GumStalker *));
+  if (ctx == NULL)
+    return _Unwind_Find_FDE (pc, bases);
+
+  real_address = gum_metal_hash_table_lookup (ctx->excluded_calls, pc + 1);
+  if (real_address == NULL)
+  {
+    real_address = gum_invocation_stack_translate (gum_interceptor_get_current_stack(),
+        (gpointer) pc + 1);
+    if (real_address == NULL)
+    {
+      real_address = gum_metal_hash_table_lookup (ctx->mappings, pc + 1);
+      if (real_address == NULL)
+      {
+        return _Unwind_Find_FDE (pc, bases);
+      }
+    }
+  }
+
+  result = _Unwind_Find_FDE (real_address - 1, bases);
+  return result;
+}
+
+static unsigned long
+gum_stalker_exception_getip (struct _Unwind_Context * context)
+{
+  unsigned long ip;
+  unsigned long result;
+  GumExecCtx * ctx;
+  gpointer real_address;
+
+  ctx = gum_stalker_get_exec_ctx (GUM_IC_GET_REPLACEMENT_DATA (
+        gum_interceptor_get_current_invocation (), GumStalker *));
+  if (ctx == NULL)
+    return _Unwind_GetIP (context);
+
+  ip = _Unwind_GetIP (context);
+
+  real_address = gum_metal_hash_table_lookup (ctx->excluded_calls, (gconstpointer) ip);
+  if (real_address == NULL)
+  {
+    real_address = gum_invocation_stack_translate (gum_interceptor_get_current_stack(),
+        (gpointer) ip);
+    if (real_address == NULL)
+    {
+      real_address = gum_metal_hash_table_lookup (ctx->mappings, (gconstpointer) ip);
```

Comment:
The values in this one are `GumExecBlock *` values, so this looks like it needs some work.

Reply:
What do I need to do?

Reply:
I don't know, but will figure it out.